Source/dafny.msbuild/DafnyVersionCheckTask.cs

```diff
@@ -0,0 +1,157 @@
+         */
+        private bool Fail(string message)
+        {
+            Log.LogError(message);
+            return false;
+        }
+
+        /**
+         * Returns the Dafny version (as printed during execution) and full path; or null, if the process does not print its version in a parseable manner.
+         */
+        private static string GetDafnyExecutableVersion(string dafnyExecutablePath)
+        {
+            using var process = new Process
+            {
+                StartInfo = {FileName = dafnyExecutablePath, CreateNoWindow = true, UseShellExecute = false, RedirectStandardOutput = true}
+            };
+            // Dafny prints its version only if given at least one input file. With an invalid file name, it will print its version, then immediately exit.
+            process.StartInfo.ArgumentList.Add("");
+            process.Start();
+            process.WaitForExit();
+
+            var output = process.StandardOutput.ReadLine() ?? "";
+            var match = new Regex("^Dafny ([0-9.]+)$").Match(output);
+            return match.Success ? match.Groups[1].Value : null;
+        }
+
+        /**
+         * Returns the HEAD commit hash of the Git repository containing the working directory, or null if an error occurred.
+         */
+        private static string GetGitRepoHeadCommitHash(string workingDir)
+        {
+            using var process = RunGitProcess(workingDir, new[] {"show-ref", "-s", "--verify", "HEAD"});
+            return process.ExitCode == 0 ? process.StandardOutput.ReadLine() : null;
+        }
+
+        /**
+         * Returns the commit hash corresponding to the given tag name, or null if an error occurred.
+         */
+        private static string GetCommitHashForGitTag(string workingDir, string tagName)
+        {
+            using var process = RunGitProcess(workingDir, new[] {"show-ref", "-s", "--verify", $"refs/tags/{tagName}"});
+            return process.ExitCode == 0 ? process.StandardOutput.ReadLine() : null;
+        }
+
+        /**
+         * Returns true if the Git repository containing the working directory is dirty (has changes in its index or work tree).
+         */
+        private static bool IsGitWorkingTreeDirty(string workingDir)
+        {
+            using var process = RunGitProcess(workingDir, new[] {"diff-index", "--quiet", "HEAD", "--"});
+            return process.ExitCode switch
+            {
+                0 => false,
+                1 => true,
+                _ => throw new Exception("`git diff-index` failed")
+            };
+        }
+
+        /**
+         * Runs Git in the given working directory and with the given arguments, and returns the Process.
+         * This assumes that `git` is in the user's PATH variable.
+         * The caller is responsible for disposing of the returned Process.
+         */
+        private static Process RunGitProcess(string workingDir, IEnumerable<string> args)
```

Comment:
Seems like you could generalize this a tiny bit more to work on any target command (perhaps including the target command as the first element of `args` instead), and then reuse it in the implementation of `GetDafnyExecutableVersion`. For that matter, `DafnyVerifyTask.VerifyDafnyFile` has this pattern too. :)

Reply:
Good idea, I've generalized this in the latest commit.